paywall/src/middlewares/walletMiddleware.js

```diff
@@ -114,7 +114,10 @@ export default function walletMiddleware({ getState, dispatch }) {
 
   return function(next) {
     // Connect to the current provider
-    walletService.connect(getState().provider)
+    // We connect once the middleware has been initialized, using setTimout (warning: fragile?)
+    setTimeout(() => {
```

Comment:
This does seem really fragile - is it guaranteed that the middleware will have initialized when `setTimeout` is called?

Reply:
Yes it is, because the entirety of redux is synchronous. Middleware are allowed to do other async stuff, but when it hooks back into redux via dispatch, the entire set of middlewares/reducers is run before the process releases control back. In this case, the `setTimeout` will simply postpone the connect call until after that synchronous loop has finished, which basically looks like this:

```
createStore is called
middleware 1 is initialized
middleware 2 is initialized
store is returned
(whatever else happens in the script, setting up React tree, etc.
...
process concludes and releases control to the event loop
--> setTimeout callback is called
```

Redux has had some pressure to change this from the React devs, but because redux supports anything, and is used in all kinds of frameworks, this will never happen, and we can rely upon it being that way until the end of time. amen.